src/sync/subscribable.hpp

```diff
@@ -0,0 +1,104 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_SUBSCRIBABLE_HPP
+#define REALM_SUBSCRIBABLE_HPP
+
+#include <cstdint>
+#include <unordered_map>
+
+namespace realm {
+
+/// Generic subscribable that allows for coarse, manual notifications
+/// from class type T.
+template <class T>
+struct Subscribable {
+    /// Token that identifies an observer.
+    /// Unsubscribes when deconstructed to
+    /// avoid dangling observers.
+    struct Token {
+        Token(Subscribable& subscribable, uint64_t token)
+        : m_subscribable(subscribable)
+        , m_token(std::make_unique<uint64_t>(token))
+        {
+        }
+        Token(Token&& other)
+        : m_subscribable(std::move(other.m_subscribable))
+        , m_token(std::move(other.m_token))
+        {
+        }
+        Token& operator=(Token&& other) {
```

Comment:
This looks like it could just be `= default`. If it can't be, it needs to be noexcept and actually return `*this`.